### PR TITLE
Add support for Hardware Design Worksheets and Integration with CheckDates program

### PR DESCRIPTION
**Pull request by maxnz**
Increment version to 2.2.1 and tag pre-release with 2.2.1-beta

Added a new type of assignment: ws (worksheet)

Added integration with a new program: CheckDates
CheckDates finds the first submission for the assignment and adds it to the log file.
CheckDates is not required for cs251tk to run. Currently CheckDates is only in the [Hardware Design Specs](https://github.com/maxnz/cs241-specs) repository.

Works on the Windows Linux Subsystem - needs to be tested on macOS and Linux.

### CODE REVIEW

**Comment by hawkrives** [comment]
Initially, I like it.

**Comment by maxnz** [comment]
Those asserts are really picky. Should be fixed now.

**Comment by hawkrives** [comment]
very <i>assert</i>ive?

**Comment by hawkrives** [comment]
Would you be open to moving the cs241-specs repo to StoDevX or another org?

File: cs251tk/toolkit/__main__.py [hunk 1]
```diff
@@ -72,6 +72,22 @@ def main():
     if date:
         logging.debug('Checking out {}'.format(date))
 
+    try:
+        with chdir(os.path.join(basedir, 'data')):
+            run(['git', 'pull', 'origin', 'master'])
```
Comment (hawkrives) [inline cs251tk/toolkit/__main__.py]: Why change this to pull on every run?

**Comment by maxnz** [comment]
Definitely. I can do that now. The only thing I ask is that I get write permissions on that one because I'm currently creating the HD specs as we make and assign the homeworks.

File: cs251tk/toolkit/__main__.py [hunk 2]
```diff
@@ -72,6 +72,22 @@ def main():
...
+
+            try:
+                with chdir(os.path.join(basedir, 'data', 'source')):
+                    run('make')
+            except FileNotFoundError:
+                if not quiet or not no_update:
+                    print("Optional add-on program CheckDates not installed.\n"
```
Comment (hawkrives) [inline cs251tk/toolkit/__main__.py]: Eventually I'd like to move this program into this python script, but that's not a blocker for merging.

File: cs251tk/toolkit/__main__.py [hunk 3]
```diff
@@ -72,6 +72,22 @@ def main():
...
+                          "Install to see first commit dates for assignments.")
+
+    except FileNotFoundError:
+        print('data directory not found')
```
Comment (hawkrives) [inline cs251tk/toolkit/__main__.py]: Print errors to stderr, not stdout, please

```suggestion
        print('data directory not found', file=sys.stderr)
```

File: setup.py [hunk 4]
```diff
@@ -6,7 +6,7 @@
 
 setup(
     name='cs251tk',
-    version='2.2.0',
+    version='2.2.1',
```
Comment (hawkrives) [inline setup.py]: This should probably go to v2.3.0, since it's a new feature and not just a bugfix.

**Comment by hawkrives** [comment]
@maxnz Absolutely

File: README.md [hunk 5]
```diff
@@ -13,13 +13,14 @@ See the ["Run the thing"](#run-the-thing) section for details.
...
 
-Prerequisites: macOS/Linux, Python 3.5+, git.
+Prerequisites: macOS/Linux/Windows Linux Subsystem, Python 3.5+, git.
 
 ##### Make the folder
 ```console
 $ mkdir cs251/
 $ cd cs251
 ```
+*(For Hardware Design use cs241 for your directory name)*
```
Comment (hawkrives) [inline README.md]:
```suggestion
> For Hardware Design, use `cs241` for your directory name.
```

File: README.md [hunk 6]
```diff
@@ -134,6 +144,7 @@ The toolkit also takes a `--record` parameter. In broad strokes, `--record` does
 
 If you're running this on something other than a lab machine, you'll want to have `gcc` installed.
 
+> The specs for Hardware Design include a program called CheckDates that adds the date and time of the first commit for each assignment. This is not required, but if you want to include it, copy the `source` directory in the [Hardware Design Specs](https://github.com/maxnz/cs241-specs) into `data/source`.
```
Comment (hawkrives) [inline README.md]:
```suggestion
> The specs for Hardware Design include a program called CheckDates that adds the date and time of the first commit for each assignment. This is not required, but if you want to include it, copy the `source` directory in the [Hardware Design Specs](https://github.com/StoDevX/cs241-specs) into `data/source`.
```

File: cs251tk/student/markdownify/markdownify.py [hunk 7]
```diff
@@ -4,15 +4,24 @@
 from collections import OrderedDict
 from .process_file import process_file
 from .find_warnings import find_warnings
+from ...common import chdir
+from cs251tk.common import run
```
Comment (hawkrives) [inline cs251tk/student/markdownify/markdownify.py]:
```suggestion
from ...common import run, chdir
```

I think we can just import both of these from the same relative import.

File: cs251tk/student/markdownify/markdownify.py [hunk 8]
```diff
@@ -4,15 +4,24 @@
...
 
 
 def markdownify(spec_id, *, username, spec, basedir, debug, interact):
     """Run a spec against the current folder"""
     try:
+        """Try to run the CheckDates program. If not installed, move on."""
+
+        with chdir(os.path.join(basedir, 'data')):
+            _, res, _ = run([os.path.join(basedir, 'data', 'CheckDates'), '-t', '-s', username, '-a', spec_id])
+        if res[:9] == "[Errno 2]":
```
Comment (hawkrives) [inline cs251tk/student/markdownify/markdownify.py]: Can we find the proper Python `Exception` class and check for it here?

Comment (hawkrives) [reply cs251tk/student/markdownify/markdownify.py]: like, a `try/except` clause thing

Comment (maxnz) [reply cs251tk/student/markdownify/markdownify.py]: Done. It's apparently an IOError.

**Comment by maxnz** [comment]
I have to give PyCharm some credit: the ctrl-click feature helped me follow all the function calls to make sense of the code. It also has spell-check included.

**Comment by maxnz** [comment]
I think that's it unless if you can think of anything.

**Comment by maxnz** [comment]
There we go. Added a little more to the "CheckDates not installed" message to direct users to the README if they are interested.

**Pull request merged**